lib/stages/frbNetworkProcess.cpp

```diff
@@ -298,5 +232,303 @@ void frbNetworkProcess::main_thread() {
+                } else {
+                    // TODO: stop the pinging thread entirely?
+                    WARN("Ping listening error: {}.", errno);
+                    break;
+                }
+            }
+            for (int s : ping_src_fd) {
+#ifdef DEBUGGING
+                char src_addr_str[INET_ADDRSTRLEN
+                                  + 1]; // Used for decoding host IP in logging statements
+#endif                                  // DEBUGGING
+                if (FD_ISSET(s, &rfds)) {
+                    sockaddr_in from; // out-param for `recv_from(2)`
+                    int reply_ping_seq = receive_ping(s, from);
+                    if (reply_ping_seq >= 0) {
+                        if (dest_by_ip.count(from.sin_addr.s_addr)) {
+                            DestIpSocketTime& src = dest_by_ip[from.sin_addr.s_addr];
+                            if (reply_ping_seq == src.ping_seq) {
+                                auto now = std::chrono::steady_clock::now();
+                                DEBUG("Received ping response from: {}. Update `last_responded`.",
+                                      inet_ntop(AF_INET, &from.sin_addr, src_addr_str,
+                                                INET_ADDRSTRLEN + 1));
+                                src.last_responded = now;
+                                if (!src.dst->live) {
+                                    INFO("Host {} is responding to pings again, mark live.",
+                                         src.dst->host);
+                                    src.dst->live = true;
+                                    src.check_delay = min_ping_frequency_;
+                                } else {
+                                    // a live host was pinged successfully, back off
+                                    // the next check
+                                    if (2 * src.check_delay <= max_ping_frequency_) {
+                                        src.check_delay *= 2;
+                                    }
+                                }
+                            } else {
+                                DEBUG(
+                                    "Ignore ping reply with old sequence number ({}) from host {}",
+                                    reply_ping_seq, src.dst->host);
+                            }
+                        } else {
+                            DEBUG("Received ping response from unknown host: {}. Ignored.",
+                                  inet_ntop(AF_INET, &from.sin_addr, src_addr_str,
+                                            INET_ADDRSTRLEN + 1));
+                        }
+                    }
+                }
+            }
+        };
+
+        // Mark node as dead if it's been too long since last response
+        if (lru_dest.dst->live) {
+            auto time_since_last_live = std::chrono::steady_clock::now() - lru_dest.last_responded;
+            if (time_since_last_live > max_ping_frequency_) {
```

Comment:
I think this should fix it.  Comparing two time points will give an answer in the base clock unit.  Which likely (system dependent) be a unit smaller than seconds, so `time_since_last_live` is a large number.  To compare times they both need to be the same `duration` unit.  This is kind of the risk of using `auto` and one of the reasons I don't always like it. 
```suggestion
            if (std::chrono::duration_cast<std::chrono::seconds>(time_since_last_live) > max_ping_frequency_) {
```